src/phoenix/server/api/types/Dimension.py

```diff
@@ -186,11 +197,66 @@ def drift_time_series(
     )  # type: ignore
     def segments_comparison(
         self,
-        primary_time_range: Optional[TimeRange] = strawberry.UNSET,
+        info: Info[Context, None],
+        primary_time_range: Optional[TimeRange] = UNSET,
     ) -> Segments:
         # TODO: Implement binning across primary and reference
 
-        return Segments(segments=[], total_counts=DatasetValues(primary_value=0, reference_value=0))
+        model = info.context.model
+        count = Count()
+        summaries = {}
+        binning_method = (
+            binning.QuantileBinning(
+                reference_series=self.dimension[REFERENCE],
+            )
+            if self.dimension.data_type is CONTINUOUS
+            else binning.CategoricalBinning()
+        )
+        for role, time_range in (
+            (PRIMARY, primary_time_range),
+            (REFERENCE, None),
+        ):
+            if (df := model[role]).empty:
+                continue
+            if time_range:
+                start, stop = row_interval_from_sorted_time_index(
+                    df.index,
+                    time_range.start,
+                    time_range.end,
+                )
+                df = df.iloc[start:stop]
+            summaries[role] = binning_method.segmented_summary(
+                self.dimension,
+                df,
+                (count,),  # type: ignore
+            )
+        segments = Segments()
+        lbound, ubound = self.dimension.min_max
+        gql_bin_factory = GqlBinFactory(
+            numeric_lbound=lbound,
+            numeric_ubound=ubound,
+        )
+        all_bins = summaries[PRIMARY].index.union(summaries[REFERENCE].index)
+        if isinstance(binning_method, binning.IntervalBinning) and binning_method.bins is not None:
+            all_bins = all_bins.union(binning_method.bins)
+        for bin in all_bins:
+            values: Dict[ms.DatasetRole, Any] = defaultdict(lambda: None)
+            for role in ms.DatasetRole:
+                if model[role].empty:
+                    continue
+                try:
+                    result = summaries[role].loc[bin]
+                except KeyError:
+                    result = {}
+                values[role] = count.get_value(result)
+            segments <<= Segment(
+                bin=gql_bin_factory(bin),
+                counts=DatasetValues(
+                    primary_value=values[PRIMARY],
+                    reference_value=values[REFERENCE],
+                ),
+            )
```

Comment:
I think while this is syntactically stylish - I worry at least for now that I would have the knowhow to find how this works. For that reason it might be worth keeping this verbose but debuggable